src/Control/Monad/Bayes/Sampler.hs

```diff
@@ -43,97 +40,52 @@ import Control.Monad.Bayes.Class
         uniform
       ),
   )
-import Control.Monad.ST (ST, runST, stToIO)
-import Control.Monad.State (State, state)
-import Control.Monad.Trans (MonadIO, lift)
-import Control.Monad.Trans.Reader (ReaderT, ask, mapReaderT, runReaderT)
+import Control.Monad.ST (ST)
+import Control.Monad.Trans.Reader (ReaderT (..), runReaderT)
 import Data.Fixed (mod')
-import Numeric.Log (Log (ln))
-import System.Random.MWC
-  ( Gen,
-    GenIO,
-    GenST,
-    Seed,
-    Variate (uniform, uniformR),
-    create,
-    createSystemRandom,
-    restore,
-    save,
-  )
+import Numeric.Log (Log (..))
 import System.Random.MWC.Distributions qualified as MWC
+import System.Random.Stateful (IOGenM, STGenM, StatefulGen, StdGen, mkStdGen, newIOGenM, newSTGenM, uniformDouble01M, uniformRM)
 
--- | An 'IO' based random sampler using the MWC-Random package.
-newtype SamplerIO a = SamplerIO (ReaderT GenIO IO a)
-  deriving newtype (Functor, Applicative, Monad, MonadIO)
-
--- | Initialize a pseudo-random number generator using randomness supplied by
--- the operating system.
--- For efficiency this operation should be applied at the very end, ideally
--- once per program.
-sampleIO :: SamplerIO a -> IO a
-sampleIO (SamplerIO m) = createSystemRandom >>= runReaderT m
-
--- Useful for reproducibility.
-sampleIOfixed :: SamplerIO a -> IO a
-sampleIOfixed (SamplerIO m) = create >>= runReaderT m
+newtype Sampler g m a = Sampler (StatefulGen g m => ReaderT g m a)
 
--- | Like 'sampleIO' but with a custom pseudo-random number generator.
-sampleIOwith :: SamplerIO a -> Gen F.RealWorld -> IO a
-sampleIOwith (SamplerIO m) = runReaderT m
+instance Functor (Sampler g m) where
+  fmap f (Sampler s) = Sampler $ fmap f s
 
-fromSamplerST :: SamplerST a -> SamplerIO a
-fromSamplerST (SamplerST m) = SamplerIO $ mapReaderT stToIO m
+instance Applicative (Sampler g m) where
+  pure x = Sampler $ pure x
+  (Sampler f) <*> (Sampler x) = Sampler $ f <*> x
 
-instance MonadSample SamplerIO where
-  random = fromSamplerST random
+runSampler :: StatefulGen g m => Sampler g m a -> ReaderT g m a
```

Comment:
Minor, but we could put this in the definition of `Sampler` as: `newtype Sampler g m a = Sampler { runSampler :: StatefulGen g m => ReaderT g m a}`

Reply:
nvm, it didn't work cause of the constraint, as you probably already knew